File: C++/Leetcode  106.cpp
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode* Helper(vector<int>& Inor, vector<int>& Post, int st1, int ed1, int st2, int ed2)
    {
        if (st1 == ed1 || st2 == ed2)
        {
            return new TreeNode(Inor[st1]);
        }

        if (st1 > ed1 || st2 >  ed2)
        {
            return NULL;
        }

        int root_value = Post[ed2];
        int index = find(Inor.begin(),Inor.end(),root_value) - Inor.begin();

        TreeNode* Root = new TreeNode(root_value);
        int left_length = index - 1 - st1;
        Root -> left = Helper(Inor, Post, st1, index-1, st2 , st2 + left_length);
        Root -> right = Helper(Inor, Post, index + 1, ed1, st2 + left_length + 1, ed2 - 1);
        return Root;
    }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return Helper(inorder, postorder, 0, inorder.size()-1, 0 , postorder.size()-1);
    }
};

File: Python/Leetcode 106.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from ast import List


class Solution:
    def Helper(self,inorder: List[int],postorder: List[int], st1, ed1, st2, ed2):
        
        if st1 == ed1 or st2 == ed2:
            return TreeNode(inorder[st1])
        
        if st1 > ed1 or st2 > ed2:
            return None
        
        root_value = postorder[ed2]
        index = inorder.index(root_value)

        Root = TreeNode(root_value)
        left_length = index - 1 - st1

        Root.left = self.Helper(inorder, postorder, st1, index - 1, st2, st2 + left_length)
        Root.right = self.Helper(inorder,postorder, index + 1, ed1, st2 + left_length + 1, ed2 - 1)

        return Root 

        
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        return self.Helper(inorder,postorder,0,len(inorder)-1,0,len(postorder)-1)
